README
This is a simple script designed to read one or more Wiktionary XML dump files
and import definitions into Redis. It needs more far rigorous testing, but 
the results seem good enough to experiment with.

Usage
=====

$ wiktionary_to_redis.pl <wiktionary_dump.xml> <2nd_wiktionary_dump.xml> ...

You can import one or multiple files. Definitions in files listed first are
used. This was designed to use definitions from Simple Wiktionary if available,
then English Wiktionary as a fallback.

Make sure Redis (http://redis.io) is running first. 
Definitions are imported as a list with format <part of speech>:::<definition>.

After the import is finished, you can access definitions like this:

$ redis-cli
redis> LRANGE sink 0 -1
1. "noun:::A '''sink''' is a [[container]] for water to wash things in, usually with a [[drain]]."
2. "noun:::A '''sink''' is a [[device]] that [[remove]]s [[heat]] or [[energy]]."
3. "verb:::{{ti verb}} If something '''sinks''', it goes down, usually into water."

Queries with spaces need to use quote marks via the Redis CLI:

redis> LRANGE "Pacific Ocean" 0 -1
1. "proper_noun:::The '''Pacific Ocean''' is a very large [[body]] of water east of [[Asia]] and west of the [[America]]s."

Getting the data
================

Dump files are available from:

http://dumps.wikimedia.org/

Currently, it's only been tested to work with the Simple English and English
sites:

http://download.wikimedia.org/simplewiktionary/latest/simplewiktionary- \
latest-pages-articles.xml.bz2
http://download.wikimedia.org/enwiktionary/latest/enwiktionary-latest- \
pages-articles.xml.bz2

Wiki markup on other Wiktionary sites varies, so I can't guarantee it will work
for anything else. By default, it will assume the same markup as English
Wiktionary.

Updating the dictionary
=======================

If you're running the script with updated dump data, current definitions won't
be updated, but new words will be added.

If you want to re-import a new dump into the dictionary, you'll need to clear
the data already in Redis to avoid duplicate definitions:

$ redis-cli
redis> flushall

Dependencies
============

Redis
MediaWiki::DumpFile
Data::Dumper


CHANGES
0.2

* Added initial support for parsing English Wiktionary dump
* Previously reading in whole file than writing, now read/write word by word

0.1

* Initial release
* Parse Simple Wiktionary dump into Redis



t/tests.pl
use Test::More;
#use AnyEvent::Redis;
use Redis;
use Data::Dumper;

my %test_words = (
    free => 6,
    about => 3,
    'pacific ocean' => 1, # phrase
    actuality => 2, 
    set => 15,
    ramparts => 1, # only in en Wikipedia
    'touch base' => 2, # only in en Wikipedia
    seoul => 1,  # capitalised in dict
    underscores => 1,
    rages => 1, # markup is {{noun|rage|rages}} - how to handle?
);

my $REDIS_SERVER_IP = '127.0.0.1:6379';
my $r;

eval{
    $r = Redis->new(server => $REDIS_SERVER_IP);
};

plan tests => scalar keys(%test_words);

for my $test_word (keys %test_words){
    print STDERR "Word:$test_word\n";
    my @vals = $r->lrange($test_word, 0, -1);
    is scalar @vals, $test_words{$test_word}, "Correct defns # for $test_word:" . @vals;
}

done_testing;


t/en_wiktionary.pl
#!/usr/bin/perl -w

use Test::More;
#use AnyEvent::Redis;
use Redis;
use Data::Dumper;

my %test_words = (
    ramparts => 1, # only in en Wikipedia
    'touch base' => 2, # only in en Wikipedia
    'antique' => 5,
    'hurts' => 3,
    'cat' => 21,
);

my $REDIS_SERVER_IP = '127.0.0.1:6379';
my $r;

eval{
    $r = Redis->new(server => $REDIS_SERVER_IP);
};

plan tests => scalar keys(%test_words);

for my $test_word (keys %test_words){
    print STDERR "Word:$test_word\n";
    my @vals = $r->lrange($test_word, 0, -1);
    is scalar @vals, $test_words{$test_word}, "Correct defns # for $test_word:" . @vals;
}

done_testing;


t/simple_wiktionary.pl
#!/usr/bin/perl -w

use Test::More;
#use AnyEvent::Redis;
use Redis;
use Data::Dumper;

my %test_words = (
    free => 6,
    about => 3,
    'pacific ocean' => 1, # phrase
    actuality => 2, 
    set => 15,
    #ramparts => 1, # only in en Wikipedia
    #'touch base' => 2, # only in en Wikipedia
    seoul => 1,  # capitalised in dict
    underscores => 1,
    rages => 1, # markup is {{noun|rage|rages}} - how to handle?
    "cell-phone" => 1, # handle redirects?
    trombones => 1, # handle plural-only entries
    linking => 1, # past participle ref
    philanthropy => 1,
);

my $REDIS_SERVER_IP = '127.0.0.1:6379';
my $r;

eval{
    $r = Redis->new(server => $REDIS_SERVER_IP);
};

plan tests => scalar keys(%test_words);

for my $test_word (keys %test_words){
    print STDERR "Word:$test_word\n";
    my @vals = $r->lrange($test_word, 0, -1);
    is scalar @vals, $test_words{$test_word}, "Correct defns # for $test_word:" . @vals;
}

done_testing;


wiktionary_to_redis.pl
#!/usr/bin/perl -w

use MediaWiki::DumpFile;
use Data::Dumper;
use Redis;
use strict;
$|++;

our $VERSION = 0.2;

################ settings as you like ###############

our $REDIS_SERVER_IP = "127.0.0.1:6379";
our $DEBUG = 0;
our $REDIRECT_LABEL = "redir";
# En Wiktionary contains multi-languages - limit to below
# only tested for English
our $TARGET_LANGUAGE = "English";

################### settings finish ################

our $REDIS = redis_connect();

our %PARSERS = (
    'http://simple.wiktionary.org/wiki/Main_Page' => \&simplewik_parse,
    'http://en.wiktionary.org/wiki/Wiktionary:Main_Page' => \&enwik_parse,
    'default' => \&enwik_parse,
);

my @article_filters = ('Wiktionary:', 'MediaWiki:', 'Help:', 'Appendix:', 
                       'Main page:', 'Category:', 'Template:', 'Index:');

my (@mediawiki_dumps) = @ARGV;

if (grep {-f $_} @mediawiki_dumps != scalar @mediawiki_dumps 
    || length(@mediawiki_dumps) == 0){
     print STDERR "Usage: $0 <wiktionary XML dumpfile 1> <dumpfile 2> ...\n\n";
     print STDERR "Definitions for earlier files used first.\n";
     print STDERR "i.e. dumpfile 2 fills in the gaps not found in dumpfile 1\n";
     exit;
}

for my $mediawiki_dump (@mediawiki_dumps){
    print STDERR "Parsing $mediawiki_dump...\n";
    read_in_xml_dict($mediawiki_dump, \@article_filters);
}

print join(", ", @mediawiki_dumps) ." now imported into Redis.\n\n";

print<<"EOI";
Find words like this:
\$ redis-cli
redis> LRANGE <word> 0 -1
redis> LRANGE table 0 -1
redis> LRANGE "Pacific Ocean" 0 -1

EOI

########## Subs ##########

sub read_in_xml_dict{
    my ($dict_file, $article_filters) = @_;

    my $mw = MediaWiki::DumpFile->new;
    my $pages = $mw->pages($dict_file);
    my $base = $pages->base;
    my $item_count = 0;

    while (defined(my $page = $pages->next)){
        my $text = $page->revision->text;
        my $word = $page->title;
        my %defns;
        my $parser = $PARSERS{$base} || $PARSERS{default};

        print STDERR "Parsed $word: $item_count\n" if ++$item_count % 1000 == 0;

        # skip non-word articles
        next if grep { $word =~ /$_/ } @$article_filters;

        # check if word already stored, skip if so
        my @stored_defs = $REDIS->lrange(lc($word), 0, -1);
        next if scalar @stored_defs > 0;
       
        %defns = &$parser($text, $page);

        if (keys %defns){
            write_to_redis($word, \%defns);
        }
        else {
            print STDERR "   ** No defns for $word\n" if $DEBUG;
        }
    }

}

sub write_to_redis{
    my ($word, $defns) = @_;

   for my $pos (sort keys %{$defns}){
       for my $defn (@{$defns->{$pos}}){
           print STDERR "  Adding $word:$pos: $defn\n" if $DEBUG;
           my $cmd = $REDIS->rpush(lc($word), "$pos\:\:\:$defn") if $defn;
       }
   }
}

sub redis_connect{
    my $r = Redis->new(server => $REDIS_SERVER_IP);
    $r->ping || die "No server";

    print STDERR "Connected to Redis...\n";
    return $r;
}

############# Parsers ################

sub simplewik_parse{
    my ($text) = @_;

    my $pos;
    my %defns = ();

    foreach my $line (split(/\n/, $text)){
        chomp $line;

        if ($line =~ /(?=^\{?)\=\= ?([\w\s]+) ?\=\=/){
            $pos = lc($1);
            $pos =~ s/^ | $//g;
            $defns{$pos} = ();
        }
        elsif ($pos && $line =~ /^\#([^:].*)/){
            my $def = $1;
            $def =~ s/^ | $//g;
            push @{$defns{$pos}}, $def;
        }
        elsif ($pos && $line =~ /\{\{plural of\|(.*?)\}\}/){
            push @{$defns{$pos}}, $line;
        }
        elsif ($line =~ /^\#REDIRECT ?\[\[(.*?)\]\]/i){
            push @{$defns{$REDIRECT_LABEL}}, $1;
        }
    }

    return %defns;
}

sub enwik_parse{
    my ($text) = @_;

    my $pos;
    my $correct_language = 1;
    my %defns = ();
    my $lang;
    my $head_depth = 3;

    foreach my $line (split(/\n/, $text)){
        chomp $line;
        # get part of speech ($pos)
        if ($line =~ /^\=\= ?([\w\s]+) ?\=\=/){
            # TODO: this only works for English, as it is always first on page
            last if $1 ne $TARGET_LANGUAGE;
        }
        # if there are multiple etymologies, headings for PoS are one lv deeper
        elsif ($line =~ /^\={$head_depth} ?Etymology 1 ?\={$head_depth}/i){
            $head_depth++;
        }
        # a part of speech header e.g. === Noun ===
        elsif ($line =~ /^\={$head_depth}[ \{\}]*([\w\s]+)[ \{\}]*\={$head_depth}/){
            $pos = lc($1);
            $pos =~ s/^ | $//g;
            $defns{$pos} = () if !exists $defns{$pos};
        }
        # the definition itself
        elsif ($pos && $line =~ /^\#([^:*].*)/){
            my $def = $1;
            $def =~ s/^ | $//g;
            push @{$defns{$pos}}, $def;
        }
    }

    return %defns;
}

